SimPEG/electromagnetics/static/induced_polarization/simulation.py

```diff
@@ -101,145 +84,24 @@ def getJtJdiag(self, m, W=None):
 
         if self.gtgdiag is None:
             J = self.getJ(m)
-
-            # Need to check if multiplying weights makes sense
             if W is None:
-                W = np.ones(J.shape[0])
+                W = self._scale ** 2
             else:
-                W = W.diagonal() ** 2
-
-            diag = np.zeros(J.shape[1])
-            for i in range(J.shape[0]):
-                diag += (W[i]) * (J[i] * J[i])
+                W = (self._scale * W.diagonal()) ** 2
 
-            self.gtgdiag = diag
+            self.gtgdiag = np.einsum("i,ij,ij->j", W, J, J)
```

Comment:
huh, that's cool. is this much more efficient?

Reply:
Much more, Dom did a few tests with it to verify in #968